### PR TITLE
Finish first draft of the ARCHITECTURE.md

### LINKED ISSUE #1779: Add architecture documentation
Providing an [architecture.md](https://matklad.github.io//2021/02/06/ARCHITECTURE.md.html) sounds like a good idea to make it easier to contribute to this project. I should write one.

### PR DESCRIPTION
**Pull request by dominikschulz**
RELEASE_NOTES=[ENHANCEMENT] Add ARCHITECTURE.md

Fixes #1779

Signed-off-by: Dominik Schulz <dominik.schulz@gauner.org>

### CODE REVIEW

File: ARCHITECTURE.md [hunk 1]
```diff
@@ -112,37 +120,113 @@ during lookup.
...
-TODO: security
+NOTE: We used to support rich terminal UIs with arrow navigation and such.
+However all existing libraries that were available without CGO were either
+abandoned or buggy on some platforms and we didn't have any capacity to fix
+them. So we had to remove support for this feature.
 
 ### `internal/queue`
 
-TODO: motivation, usage
+The `queue` package implements a FIFO queue that executed strictly sequential
```
Comment (AnomalRoil) [inline ARCHITECTURE.md:149]: "that executed strictly sequential" isn't proper, not sure what you meant?

File: ARCHITECTURE.md [hunk 2]
```diff
@@ -112,37 +120,113 @@ during lookup.
...
+in the background. This allows for certain operations, like a git push, to be
+taken out of the critical path wrt. user interactions. The queue will be fully
+processed before the process exits.
+
+### `internal/store/root`
 
-### `internal/store`
+The `root store` package implements an internal password store API that (only)
+supports mounting `leaf` stores. It will forward (almost) all operations to
+it's `leaf` stores (cross stores moves being a notable exception) and do the
```
Comment (AnomalRoil) [inline ARCHITECTURE.md:158]: "it's" -> "its" `leaf` stores.   
"cross stores moves " -> "moves across stores"

File: ARCHITECTURE.md [hunk 3]
```diff
@@ -112,37 +120,113 @@ during lookup.
...
+necessary manipulation of the affected path components.
```
Comment (AnomalRoil) [inline ARCHITECTURE.md:159]: "necessary manipulation of the affected path components" --> I have no idea what this means, could you hint at it with an example or maybe rephrase?

Also you probably want a plural "manipulations" there.

File: ARCHITECTURE.md [hunk 4]
```diff
@@ -112,37 +120,113 @@ during lookup.
...
 
-TODO: everything
+This package makes `gopass` multi-store capable.
+
+### `internal/store/leaf`
+
+The `leaf store` package implements a password store that is mostly compatible
+with any other password store implementation (while aiming for interoperability,
+not 100% feature parity). The low-level operations like filesystem and / or
```
Comment (AnomalRoil) [inline ARCHITECTURE.md:167]: "not 100% feature parity" is missing a verb.

File: ARCHITECTURE.md [hunk 5]
```diff
@@ -112,37 +120,113 @@ during lookup.
...
+version control and crypto operations are passed to the configured `storage`
+or `crypto` backend.
 
 ### `internal/tree`
 
-TODO: why
+The `tree` package implements a simple tree structure that can be printed
+similar to the output of the Unix tool `tree`. It does support different
```
Comment (AnomalRoil) [inline ARCHITECTURE.md:174]: "that can be printed similar to the output" --> rephrase. Maybe "that prints an output similar to this of ..."

File: ARCHITECTURE.md [hunk 6]
```diff
@@ -112,37 +120,113 @@ during lookup.
...
+`gopass` specific properties (like mounts or templates) not easily implemented
+with other tree packages.
 
 ### `internal/updater`
 
-TODO: security
+The `updater` package implements a secure and anonymous self updater.
+
+Note: The self updater contacts GitHub. If that is concern one should use
```
Comment (AnomalRoil) [inline ARCHITECTURE.md:182]: "If that is concern " --> "If this is a concern "

File: ARCHITECTURE.md [hunk 7]
```diff
@@ -112,37 +120,113 @@ during lookup.
...
+other sources, e.g. distro packages.
+
+It retrieves the latest stable release from GitHub, fetches it's metadata
```
Comment (AnomalRoil) [inline ARCHITECTURE.md:185]: "it's metadata" --> its metadata

File: ARCHITECTURE.md [hunk 8]
```diff
@@ -112,37 +120,113 @@ during lookup.
...
+and refuse to update these.
 
-### `pkg/...`
+### `pkg/`
 
-TODO: everything
+The package `pkg/` contains our public API surface, i.e. packages we want or
+have to expose externally. Some packages (e.g. `otp`) are only exposed because
+they are being used by some of our integrations. Others (e.g. `pinentry` or
+`pwgen`) are designed for wider use.
```
Comment (AnomalRoil) [inline ARCHITECTURE.md:196]: Do we have plan to split the wider use packages into their own repo at some point?
Wouldn't that be better?

File: ARCHITECTURE.md [hunk 9]
```diff
@@ -112,37 +120,113 @@ during lookup.
...
+
+#### `pkg/appdir`
+
+The `appdir` package contains a set of [XDG Base Directory Specification](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html)
+compatible implementation with some `gopass` specifics. For testing purposes
```
Comment (AnomalRoil) [inline ARCHITECTURE.md:201]: implementationS

File: ARCHITECTURE.md [hunk 10]
```diff
@@ -112,37 +120,113 @@ during lookup.
...
+we want to honor the setting of `GOPASS_HOMEDIR` before everything else, so our
+implementation has to take this into account before following the XDG spec.
+
+#### `pkg/clipboard`
+
+The `clipboard` package is a wrapper around a clipboard package that adds
+support for clearing the clipboard.
+
+#### `pkg/ctxutil`
```
Comment (AnomalRoil) [inline ARCHITECTURE.md:210]: Isn't this going away?

File: ARCHITECTURE.md [hunk 11]
```diff
@@ -112,37 +120,113 @@ during lookup.
...
+This package is designed as the main entry point for any integration that wants
+to integrate with gopass.
 
 ### `tests`
 
-TODO: integration tests
+`gopass` comes with a comprehensive set of integration tests, i.e. tests that
+are executed by running a newly compiled gopass binary without access to any
+kind of internal state. These tests can't be as exhaustive as the unit tests
+but they exist to ensure basic functionality isn't broken by a change.
```
Comment (AnomalRoil) [inline ARCHITECTURE.md:230]: --> basic functionalities aren't

**Review by AnomalRoil — CHANGES_REQUESTED** [review]

File: ARCHITECTURE.md (hunk 1 above)
Comment (dominikschulz) [reply ARCHITECTURE.md:149]: that was redundant. removed.

File: ARCHITECTURE.md (hunk 8 above)
Comment (dominikschulz) [reply ARCHITECTURE.md:196]: it would be cleaner but also more annoying. I'd still consider it an option, but not a promise.

File: ARCHITECTURE.md (hunk 10 above)
Comment (dominikschulz) [reply ARCHITECTURE.md:210]: Maybe. Not sure if/when we'll be there.
For now one should be aware of the `ctxutil` package.

**Comment by dominikschulz** [comment]
Thanks a lot!
